crates/schemalate/src/firebolt/firebolt_schema_builder.rs

```diff
@@ -79,17 +82,37 @@ pub fn build_firebolt_schema(binding: &Binding) -> Result<TableSchema, Error> {
         let is_key = fs.keys.contains(field);
         let (shape, exists) = schema_shape.locate(&Pointer::from_str(&projection.ptr));
 
-        let fb_type = projection_type_to_firebolt_type(shape).ok_or(Error::UnknownType {
-            r#type: shape.type_.to_string(),
-            field: field.clone(),
-        })?;
+        let mut nullable = !exists.must();
+        let fb_type;
+        if exists == Implicit && !shape.type_.is_single_type() {
+            // If there's no specific type in schema, try to infer it from projection
+            let inferred_type = projection.inference.as_ref().ok_or(Error::UnknownType {
+                r#type: shape.type_.to_json_array(),
+                field: field.clone(),
+            })?;
+
+            fb_type = projection_implicit_type_to_firebolt_type(inferred_type).ok_or(
+                Error::UnknownType {
+                    r#type: inferred_type.types.join(","),
+                    field: field.clone(),
+                },
+            )?;
+        } else {
+            fb_type = projection_type_to_firebolt_type(shape).ok_or(Error::UnknownType {
+                r#type: shape.type_.to_string(),
+                field: field.clone(),
+            })?;
+
+            nullable = nullable || shape.type_.overlaps(types::NULL);
+        }
```

Comment:
I think this piece can be refactored so that we use a single function for converting projection type to firebolt type, it's just that in one scenario we first find the inferred type and pass it to the same function.

Is there a reason we need the two functions to have different behaviour?

Reply:
There were two reasons to create a separate function here:
* `projection_type_to_firebolt_type` operates on a `Shape` object, which contains more information than `Inference`, which only holds type names as a vector of strings.
* Since `Shape` is more complex we're able to get the inner array type, while `Inference` does not contain it (as far as I understand) and would only have "array" as a type.